wasm/validator/src/recursive.rs

```diff
@@ -0,0 +1,290 @@
+//! Module with structs and functions for recursive validation.
+
+#![deny(clippy::wildcard_enum_match_arm)]
+
+use iroha_wasm::data_model::{
+    isi::{DirectInstruction, EitherInstruction, IndirectInstruction},
+    transaction::TransactionInfo,
+};
+
+use super::*;
+
+/// Validate operation recursively with given validation functions.
+///
+/// This function should be your validator entrypoint in **99%** of cases.
+///
+/// # Transaction validation
+///
+/// Transaction validation is done with 2 steps:
+///
+/// 1. Extract [`TransactionInfo`](TransactionInfo) from [`Transaction`]
+/// and pass it to `validate_transaction` function;
+/// 2. Validate every instruction in transaction, see next section.
```

Comment:
IMHO the assumption that transaction validation can be split into two independent steps is incorrect. This imposes an unnecessary limitation on the number of validators you can write. For instance, a user might want to write a validator that looks into metadata and use the information stored there to validate instructions in this transactions. Therefore, **transaction validation cannot be split into subparts**. 

An added benefit is that you won't need `TransactionInfo` in the data model

P.S I understand that the user can roll out a completely custom validator to get over this limitation but the current API is setting up a precedent we'd rather not follow.

Reply:
To add up, current API is favoring the case where transaction info and instructions are validated independently. We don't have people writing custom validators, but my gut feeling is that current API will largely be useless because it's favoring the wrong use case. The user would more often than not combine validation of transaction info and instructions

Reply:
Okay, that's easy to fix, I can just pass `TransactionInfo` as another parameter to `validate_instruction`. But I still think we need `TransactionInfo` it-self

Reply:
I would have hoped that I have made a convincing case of why splitting transaction validation (and having `TransactionInfo`) is redundant at best and actually detrimental at worst

Reply:
@mversic , updated, check it out